File: exercise/C#/day23/ControlSystem.External/Reindeer.cs
namespace ControlSystem.External
{
    public class Reindeer
    {
        private int _spirit = 0;
        private int _age = 0;
        private readonly string _name;
        public bool Sick = false;
        public int TimesHarnessing = 0;
        private int _powerPullLimit = 0;

        public Reindeer(string name, int age, int spirit)
            : this(name, age, spirit, false)
        {
        }

        public Reindeer(string name, int age, int spirit, bool sick)
        {
            _name = name;
            _spirit = spirit;
            _age = age;
            Sick = sick;

            _powerPullLimit = age <= 5 ? 5 : 5 - (age - 5);
        }

        public float GetMagicPower()
        {
            if (!Sick || NeedsRest())
            {
                if (_age == 1)
                    return _spirit * 0.5f;
                else if (_age <= 5)
                    return _spirit;
                else
                    return _spirit * 0.25f;
            }
            else
            {
                return 0;
            }
        }

        public bool NeedsRest()
        {
            if (!Sick)
            {
                return TimesHarnessing == _powerPullLimit;
            }
            else
            {
                return true;
            }
        }
    }
}

File: exercise/ts/day23/src/external/deer/reindeer.ts
export class Reindeer {
    private spirit = 0;
    private age = 0;
    private name: string;
    public sick = false;
    public timesHarnessing = 0;
    private powerPullLimit = 0;

    constructor(name: string, age: number, spirit: number, sick: boolean = false) {
        this.name = name;
        this.spirit = spirit;
        this.age = age;
        this.sick = sick;
        this.powerPullLimit = age <= 5 ? 5 : 5 - (age - 5);
    }

    public getMagicPower(): number {
        if (!this.sick || this.needsRest()) {
            if (this.age === 1) return this.spirit * 0.5;
            if (this.age <= 5) return this.spirit;
            return this.spirit * 0.25;
        }
        return 0;
    }

    public needsRest(): boolean {
        if (!this.sick) {
            return this.timesHarnessing === this.powerPullLimit;
        }
        return true;
    }
}
